pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,339 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	networking "istio.io/api/networking/v1alpha3"
+
+	"github.com/yl2chen/cidranger"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/protocol"
+	"istio.io/pkg/log"
+)
+
+var (
+	notReadyEndpointkey                        = "networking.alpha.istio.io/notReadyEndpoints"
+	_                   model.Controller       = &MCPDiscovery{}
+	_                   model.ServiceDiscovery = &MCPDiscovery{}
+)
+
+// namedRangerEntry for holding network's CIDR and name
+type namedRangerEntry struct {
+	name    string
+	network net.IPNet
+}
+
+// returns the IPNet for the network
+func (n namedRangerEntry) Network() net.IPNet {
+	return n.network
+}
+
+// DiscoveryOptions stores the configurable attributes of a Control
+type DiscoveryOptions struct {
+	XDSUpdater   model.XDSUpdater
+	Env          *model.Environment
+	ClusterID    string
+	DomainSuffix string
+}
+
+func (o *DiscoveryOptions) mixerEnabled() bool {
+	return o.Env != nil && o.Env.Mesh != nil && (o.Env.Mesh.MixerCheckServer != "" || o.Env.Mesh.MixerReportServer != "")
+}
+
+// MCPDiscovery provides storage for NotReadyEndpoints
+type MCPDiscovery struct {
+	// [ip:port]config
+	notReadyEndpoints   map[string]*model.Config
+	notReadyEndpointsMu sync.RWMutex
+	options             *DiscoveryOptions
+	networkForRegistry  string
+	ranger              cidranger.Ranger
+}
+
+// NewMCPDiscovery provides a new instance of Discovery
+func NewMCPDiscovery(options *DiscoveryOptions) *MCPDiscovery {
+	return &MCPDiscovery{
+		notReadyEndpoints: make(map[string]*model.Config),
+		options:           options,
+	}
+}
+
+// InitNetworkLookup will read the mesh networks configuration from the environment
+// and initialize CIDR rangers for an efficient network lookup when needed
+func (d *MCPDiscovery) InitNetworkLookup(meshNetworks *meshconfig.MeshNetworks) {
```

Comment:
Why does it have to be here ? This should be in the eds or model, not in an individual registry.